pallets/stake/src/lib.rs

```diff
@@ -0,0 +1,473 @@
+            Self::update_ledger(&seller, &seller_ledger);
+
+            // TODO transfer the right of use to buyer
+            Ok(())
+        }
+
+        pub fn remove_order_or_expire_deal(seller: T::AccountId, deal_id: DealId<T>) -> DispatchResult {
+            <Users<T>>::mutate(&seller, |ledger| {
+                ledger.active += ledger.seller[&deal_id];
+                ledger.seller.remove(&deal_id);
+            });
+            Ok(())
+        }
+
+        /// the unique hash for a report in a span of time
+        fn report_sp_id<O: AutomataOffence<T::AccountId>>(
+            special_id: &GeodeIdd<T::AccountId>,
+            offender: &T::AccountId,
+        ) -> ReportIdOf<T> {
+            let unique_time = Self::time_span_filter(&special_id);
+            (O::ID, unique_time.encode(), offender).using_encoded(T::Hashing::hash)
+        }
+
+        /// Divide time into spans.
+        fn time_span_filter(time: &GeodeIdd<T::AccountId>) -> u64 {
+            time.offline_time / 100
+        }
+
+        /// if the offence is reported, return true.
+        fn duplicate(id: ReportIdOf<T>) -> bool {
+            <Reports<T>>::contains_key(id)
+        }
+
+        /// compute slash for an offence
+        fn compute_slash(
+            balance_staking: &BalanceOf<T>,
+            extra_slash_rate: &Perbill,
+            reward_rate: &Perbill,
+        ) -> (BalanceOf<T>, BalanceOf<T>, BalanceOf<T>) {
+            let mut compensation_to_user: BalanceOf<T> = Zero::zero();
+            let mut val_extra_slashed: BalanceOf<T> = Zero::zero();
+
+            let to_extra_slash: BalanceOf<T> = *extra_slash_rate * *balance_staking;
+
+            val_extra_slashed += to_extra_slash;
+
+            let to_user: BalanceOf<T> = *reward_rate * to_extra_slash;
+            compensation_to_user += to_user;
+
+            let reward_payout_to_reporter: BalanceOf<T> = val_extra_slashed - compensation_to_user;
+
+            (
+                val_extra_slashed,
+                compensation_to_user,
+                reward_payout_to_reporter,
+            )
+        }
+
+        /// slash the offence,reward the reporter.
+        fn do_slash_reward_compensate(
+            provider: &T::AccountId,
+            offence_geoid: &T::AccountId,
+            user: &T::AccountId,
+            reporters: &[T::AccountId],
+            val_slashed: BalanceOf<T>,
+            val_reward: BalanceOf<T>,
+            val_compensation: BalanceOf<T>,
+        ) {
+            //TODO:: Return fee
+
+            // Provider slash
+            let mut ledger: UserLedgerOf<T> = <Users<T>>::get(provider);
+            ledger.total -= val_slashed;
+            let new_stake = *ledger.seller.get(offence_geoid).unwrap() - val_slashed;
+            ledger.seller.insert(offence_geoid.clone(), new_stake);
+            <Users<T>>::insert(provider, ledger);
+
+            // reward report
+            //TODO:: div the val_reward
```

Comment:
Can implement the division here? The mod value can be added to the user compensation.

Reply:
The pr just upgrade stake pallet from frame v1 to v2. Could we use other pr to implement dividing, I am not clear about the dividing and how to implement it now.

Reply:
Sure no problem